### PR TITLE
[SPARK-11140][CORE] Transfer files using network lib when using NettyRpcEnv - 1.6.version.

### PR DESCRIPTION
**Pull request by vanzin**
This patch is the same code as in SPARK-11140 in master, but with some added code to still use the HTTP file server by default in NettyRpcEnv. This is mostly to avoid conflicts when backporting patches to 1.6.

### CODE REVIEW

File: docs/security.md [hunk 1]
```diff
@@ -157,7 +158,7 @@ configure those ports.
     <td>(random)</td>
     <td>File server for files and jars</td>
     <td><code>spark.fileserver.port</code></td>
-    <td>Jetty-based</td>
```
Comment (zsxwing) [inline docs/security.md]: Could you revert this change?

File: docs/configuration.md [hunk 2]
```diff
@@ -1027,6 +1028,7 @@ Apart from these, the following properties are also available, and may be useful
   <td>(random)</td>
   <td>
     Port for the driver's HTTP file server to listen on.
+    This is only relevant when using the Akka RPC backend.
```
Comment (zsxwing) [inline docs/configuration.md]: nit: should be removed

**Comment by vanzin** [comment]
I'll revert all the doc changes.

File: core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala [hunk 3]
```diff
@@ -300,6 +323,132 @@ private[netty] class NettyRpcEnv(
...
+
+    override def onFailure(streamId: String, cause: Throwable): Unit = {
+      logError(s"Error downloading stream $streamId.", cause)
+      source.setError(cause)
+      sink.close()
+    }
+
+  }
+
+  private class HttpBasedFileServer extends RpcEnvFileServer {
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala]: Could you move this one to a new file? So that we don't need to resolve the conflicts when adding new codes at the end of NettyRpcEnv.

**Comment by zsxwing** [comment]
LGTM

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
retest this please

**Comment by rxin** [comment]
I've merged this. 

@vanzin  can you close the pull request?
